### ISSUE TITLE
Adding support for raw packets

### ISSUE THREAD

**Issue by seirdb** [tokio-rs/tokio#203]
I'm working on a project where I need access to raw Layer 2 packets. IMHO, it would be nice if there was support for raw packets (AF_PACKET e.a.) in `mio`/`tokio` directly.

In unix-like systems, a raw packet socket can easily be obtained, and will work nicely with `epoll`. On Windows, I think we'd have to go through `winpcap`.

What would be needed, besides the obvious `bind`/`send`/`receive` is support for `setsockopt` and ideally `control message` data along side `sendmsg`/`recvmsg` as well.

I have some time to spend and am willing to work towards a PR/external crate/...

So my question is, where should this live, `mio`/`tokio`/new crate/...? Any general thoughts/feedback/pointers/...?

**Comment by carllerche** [comment]
Thanks for the thought.

So, if it works with epoll, then it should integrate nicely with the Tokio stack. There isn't anything special about TCP / UDP. In fact, those implementations are being moved into a separate crate. See #224.

Given this, it would make most sense for something like this to start out in a stand alone git repo / crate and once it matures, then we can consider moving it into the Tokio project.

Now, my main concern is Windows. I don't know much about `winpcap`, but Mio uses IOCP. So, for windows support to fit in well with Tokio, it would have to work with IOCP somehow.

**Comment by seirdb** [comment]
Agreed, I'll start off with a standalone crate. I'll check out the work done in #224 as a starting point.

As far as Windows support, I personally don't have any experience with developing for it. My initial focus would be on linux.

**Comment by mrmonday** [comment]
You might like to look into [libpnet](https://github.com/libpnet/libpnet) - a lot of the work is already done, but it's lacking async/tokio support.
